### ISSUE TITLE
Turn `Subscription` into a Trait

### ISSUE THREAD

**Issue by zslayton** [zslayton/stomp-rs#46]
cc @fribeiro1 

Possibly the greatest shortcoming in the `stomp` library is that the `on_message` callback is a bare function (`fn`). This burdens the message handler with creating its own resources each time a message is received. It is impossible[1], for example, to have the message handling function send a message somewhere meaningful via a `channel` or to keep count of how many times it has been called. This dramatically limits what can be done using this crate.

There are two possible solutions to this issue:

1. Use closures instead of bare functions
2. Create a `Subscription` trait that the developer can implement.

Having used the Stomp protocol in the past mostly from dynamic languages like JS and Python, closures seemed to me like the most ergonomic way of offering this functionality. (Note the existence of issue #33.) This would look something like:

```rust 
  let topic = "/topic/messages";
  let mut message_count = 0u64;
  session.subscribe(topic, Client, |frame: &Frame| {
    message_count += 1;
    println!("Received {} messages", message_count);
  });
```
However, it may be more idiomatic to convert `Subscription` into a Trait and require the developer to create a struct with the desired environment which implements it. That would look something like:
```rust
struct ExampleSubscription {
  count: u64
}

impl Subscription for ExampleSubscription {
  fn on_message(&mut self) -> AckOrNack {
    self.count += 1;
    Ack
  }
}

fn main() {
  let subscription = ExampleSubscription::new();
  session.subscribe(topic, Client, subscription);
}
```

This is obviously more verbose and isn't quite as nice for short demonstrations. However, it feels as though it may be more fitting for non-trivial programs relying on an MQ broker.

I'm currently leaning towards the Trait approach, but I wanted to open this issue to request input on the topic before jumping into revisions.

[1] It is possible to do something like this using static data structures, but this is not idiomatic Rust.

**Comment by fribeiro1** [comment]
Perhaps because I am not a Rustecean, I like the closure approach better.

**Comment by zslayton** [comment]
This weekend I wrote two versions of the Subscription system: one that requests a closure and one that requests a `MessageHandler`, a trait requiring `fn on_message(&mut self, &Frame) -> AckOrNack`. 

The closure version certainly worked, but unfortunately the resulting API wasn't very friendly. I had to go through several attempts to write a simple client, fighting with the (temporary?) syntax for unboxed closures. The types got pretty ugly. Something like: `(&'a mut (FnMut(&Frame) -> AckOrNack) + 'b)` started to get pretty common.

I'm planning to go forward with the `MessageHandler` version of things. My hope is that I can write a separate `ToMessageHandler` trait at some point and implement it for `Fn`, `FnMut`, and `MessageHandler` so that the developer can pass any one of those three values in to the `subscribe` method.
